Sources/NIOHTTPCompression/HTTPDecompression.swift

```diff
@@ -0,0 +1,147 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2019 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import CNIOExtrasZlib
+import NIO
+
+public enum NIOHTTPDecompression {
+    /// Specifies how to limit decompression inflation.
+    public struct DecompressionLimit {
+        private enum Limit {
+            case none
+            case size(Int)
+            case ratio(Int)
+        }
+
+        private var limit: Limit
+
+        /// No limit will be set.
+        public static let none = DecompressionLimit(limit: .none)
+        /// Limit will be set on the request body size.
+        public static func size(_ value: Int) -> DecompressionLimit { return DecompressionLimit(limit: .size(value)) }
+        /// Limit will be set on a ratio between compressed body size and decompressed result.
+        public static func ratio(_ value: Int) -> DecompressionLimit { return DecompressionLimit(limit: .ratio(value)) }
+
+        func exceeded(compressed: Int, decompressed: Int) -> Bool {
+            switch self.limit {
+            case .none:
+                return false
+            case .size(let allowed):
+                return compressed > allowed
+            case .ratio(let ratio):
+                return decompressed > compressed * ratio
+            }
+        }
+    }
+
+    public enum DecompressionError: Error {
+        case limit
+        case inflationError(Int)
+        case initializationError(Int)
+    }
+
+    enum CompressionAlgorithm: String {
+        case gzip
+        case deflate
+
+        init?(header: String?) {
+            switch header {
+            case .some("gzip"):
+                self = .gzip
+            case .some("deflate"):
+                self = .deflate
+            default:
+                return nil
+            }
+        }
+
+        var window: CInt {
+            switch self {
+            case .deflate:
+                return 15
+            case .gzip:
+                return 15 + 16
+            }
+        }
+    }
+}
+
+struct Decompressor {
+    private let limit: NIOHTTPDecompression.DecompressionLimit
+    private var stream = z_stream()
+    private var inflated = 0
+
+    init(limit: NIOHTTPDecompression.DecompressionLimit) {
+        self.limit = limit
+    }
+
+    mutating func decompress(part: inout ByteBuffer, buffer: inout ByteBuffer, originalLength: Int) throws {
+        try self.stream.inflatePart(input: &part, output: &buffer)
+        self.inflated += buffer.readableBytes
```

Comment:
This doesn't work the way you want if `buffer` already has readable bytes. With general code we need to keep track of what the original number of readable bytes was.

Reply:
good catch, fixed